src/core/net/udp6.hpp

```diff
@@ -123,6 +123,10 @@ class Udp : public InstanceLocator
         {
             mHandler(mContext, &aMessage, &aMessageInfo);
         }
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+        bool mBoundToThreadStack;
```

Comment:
Yes. Can't pass checks without this. 

The problem was on Backbone Routers. Since a Backbone Router has two Udp Sockets both bound to port 61631, it does not know which one is the correct receiver for a received UDP message. This flag thus help Backbone Router to skip the Udp sockets bound to Backbone interface that are not supposed to process any UDP messages received from the Thread stack.

Reply:
This adds the overhead to all sockets. As `MessageInfo::mIstHostInterface` tells whether this message comes from Thread network interface, is it possible to add special handling for the backbone tmf port here?

Reply:
This adds the overhead to all sockets, **but only on Backbone Routers**. Note that Backbone Routers are usually way more powerful than normal SOC devices. 

The problem is in the UDP socket matching where UDP sockets know nothing about up layer COAP objects. It does not know which Udp Socket should receive a UDP message if they share the same port. We can add special handling to skip the TMF socket by comparing Socket object references. 

However I think we should not keep adding hardcoding and special handling for Netif binding issues, but to accept that OT core should be aware of different network interfaces (at least on Backbone Routers).

Reply:
Since this field is only added into the internal class, I'm wondering if this will affect upper layer users who don't have such internal classes from C API. `otUdpSocket` is directly used in the CLI library, will it cause issues when being cast to the internal `SocketHandle`?

Reply:
This is a good point. Adding `bool mBoundToThreadStack` to `otUdpSocket` brings unnecessary cost that should be avoided.

Reply:
How about managing sockets segment by segment.

```bash

socket1 -> socket2 - > ... -> socket[n] -> socket[n+1] -> ...
     ^                            ^
     |                            |
Head of all sockets(not bound)  First backbone link socket
```

Reply:
segmented linked list implemented. Please check.